### PR TITLE
Refactor codebase into Typescript

### PR DESCRIPTION
**Pull request by raygesualdo**
Steps taken to accomplish this:

- [x] Update testing and build tooling dependencies to latest versions
- [x] Add Typescript tooling dependencies
- [x] Convert JS files to TS and add type annotations
- [x] Add npm script to generate type files
- [x] Add type checks to CI builds

### CODE REVIEW

**Review by devproivansurzhenko — CHANGES_REQUESTED** [review]

File: src/pushex.spec.js [hunk 1]
```diff
@@ -148,11 +148,11 @@ describe("unsubscribe", () => {
     const sub = pushex.subscribe("chName")
 
     expect(pushex.subscriptions.chName).not.toEqual(undefined)
-    expect(sub.channel).not.toEqual(null)
+    expect(sub.channel).not.toEqual(undefined)
 
     pushex.unsubscribe("chName").then(() => {
       expect(pushex.subscriptions).toEqual({})
-      expect(sub.channel).toEqual(null)
+      expect(sub.channel).toEqual(undefined)
```
[the remark this answers is not in the document]
Comment (raygesualdo) [reply src/pushex.spec.js:155]: Correct me if I'm wrong @sb8244, but I think Steve prefers `.toEqual` with the value over the other built-in methods.

File: src/pushex.ts [hunk 2]
```diff
@@ -3,12 +3,32 @@ import { Subscription } from "./subscription"
...
+      onConnect,
+      onConnectionError,
+      socketReconnectAlgorithm
+    }: {
+      getParams: () => Promise<Record<string, any>>
+      onConnect: (instance: Pushex) => void
+      onConnectionError: (instance: Pushex) => void
+      socketReconnectAlgorithm: typeof DEFAULT_SOCKET_RECONNECT_ALGORITHM
+    }
+  ) {
```
[the remark this answers is not in the document]
Comment (raygesualdo) [reply src/pushex.ts:31]: You mean the constructor arguments type?

[the remark this answers is not in the document]
Comment (devproivansurzhenko) [reply src/pushex.ts:31]: Not just arguments. Something like that:

```
interface IPushexConfig {
  reconnectAlgorithm?: typeof DEFAULT_SOCKET_RECONNECT_ALGORITHM
  getParams?: () => Promise<Record<string, any>>
  onConnect?: (instance: Pushex) => void
  onConnectionError?: (instance: Pushex) => void
}

...

const DEFAULT_CONFIG = ...

...

export class Pushex {
  private config: IPushexConfig = DEFAULT_CONFIG

  constructor(url: string, config: IPushexConfig) {
    this.config = {...DEFAULT_CONFIG, ...config}
    ...
```

[the remark this answers is not in the document]
Comment (raygesualdo) [reply src/pushex.ts:31]: Yeah, I can could do that. Any preference @sb8244?

File: src/pushex.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sb8244) [reply src/pushex.spec.js:155]: I prefer the toEqual functions.

File: src/pushex.ts [hunk 3]
```diff
@@ -3,12 +3,32 @@ import { Subscription } from "./subscription"
...
 
-const DEFAULT_SOCKET_RECONNECT_ALGORITHM = tries => {
+const DEFAULT_SOCKET_RECONNECT_ALGORITHM = (tries: number) => {
   return [3000, 6000, 10000, 20000][tries - 1] || 30000
 }
 
 export class Pushex {
-  constructor(url, { getParams, onConnect, onConnectionError, socketReconnectAlgorithm }) {
+  private reconnectAlgorithm: typeof DEFAULT_SOCKET_RECONNECT_ALGORITHM
+  private getParams: () => Promise<Record<string, any>>
```
Comment (sb8244) [inline src/pushex.ts:12]: TIL about Record

**Comment by sb8244** [comment]
I'm going to move this over to a typescript branch to finish off

**Comment by sb8244** [comment]
Looks greaT!!!

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export { Pushex } from "./pushex"
+export { SubscriptionFunction } from "./subscription"
```
[the remark this answers is not in the document]
Comment (sb8244) [reply src/index.ts:2]: nvm, things blow up 🙃

File: src/pushex.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sb8244) [reply src/pushex.ts:31]: Refactored to an interface. I handled the default a bit different though

**Pull request merged**